cores/esp8266/WString.h

```diff
@@ -250,27 +250,29 @@ class String {
             uint16_t cap;
             uint16_t len;
         };
-
-        // SSO is handled by checking the last byte of sso_buff.
-        // When not in SSO mode, that byte is set to 0xff, while when in SSO mode it is always 0x00 (so it can serve as the string terminator as well as a flag)
-        // This allows strings up up to 12 (11 + \0 termination) without any extra space.
-        enum { SSOSIZE = sizeof(struct _ptr) + 4 }; // Characters to allocate space for SSO, must be 12 or more
+        // This allows strings up up to 11 (10 + \0 termination) without any extra space.
+        enum { SSOSIZE = sizeof(struct _ptr) + 4 - 1 }; // Characters to allocate space for SSO, must be 12 or more
+        struct _sso {
+            char     buff[SSOSIZE];
+            unsigned len  : 7;
+            unsigned flag : 1;
+        };
         enum { CAPACITY_MAX = 65535 }; // If size of capacity changed, be sure to update this enum
         union {
             struct _ptr ptr;
-            char sso_buf[SSOSIZE];
+            struct _sso sso;
         };
         // Accessor functions
-        inline bool sso() const { return sso_buf[SSOSIZE - 1] == 0; }
-        inline unsigned int len() const { return sso() ? strlen(sso_buf) : ptr.len; }
-        inline unsigned int capacity() const { return sso() ? SSOSIZE - 1 : ptr.cap; }
-        inline void setSSO(bool sso) { sso_buf[SSOSIZE - 1] = sso ? 0x00 : 0xff; }
-        inline void setLen(int len) { if (!sso()) ptr.len = len; }
-        inline void setCapacity(int cap) { if (!sso()) ptr.cap = cap; }
-	inline void setBuffer(char *buff) { if (!sso()) ptr.buff = buff; }
+        inline bool isSSO() const { return sso.flag; }
+        inline unsigned int len() const { return isSSO() ? sso.len : ptr.len; }
+        inline unsigned int capacity() const { return isSSO() ? (unsigned int)SSOSIZE - 1 : ptr.cap; }
```

Comment:
Please add comment that capacity means number of chars not including the null term.

Questions:
1. if we no longer care about zeroes in the string, shouldn't we treat the null term as just another byte? Or is that already the case, and here it's a matter of definition for capacity? 
2. Does ptr.cap mean the same, i. e. does NOT include the null term? Or is the - 1 missing in that case?

Reply:
Comment added, but it's always been kind of funky this way (old Arduino code)
1) Nul term still added at end so things like `::c_str()` can work properly.  It's just that internally it's just like a Pascal string with `\0` being as valid a character as `a`
2) capacity() is the space available for chars not including the required ending NUL, yes.  On SSO, that's 1 less than the buffer, on non-SSO that's set to 1 less than the malloc'd size.